sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriteFiles.scala

```diff
@@ -53,13 +59,17 @@ case class WriteFiles(child: LogicalPlan) extends UnaryNode {
 /**
  * Responsible for writing files.
  */
-case class WriteFilesExec(child: SparkPlan) extends UnaryExecNode {
+case class WriteFilesExec(
+    child: SparkPlan,
+    fileFormat: FileFormat,
+    partitionColumns: Seq[Attribute],
+    bucketSpec: Option[BucketSpec],
+    options: Map[String, String],
+    staticPartitions: TablePartitionSpec) extends UnaryExecNode {
   override def output: Seq[Attribute] = Seq.empty
 
-  override protected def doExecuteWrite(writeSpec: WriteSpec): RDD[WriterCommitMessage] = {
-    assert(writeSpec.isInstanceOf[WriteFilesSpec])
-    val writeFilesSpec: WriteFilesSpec = writeSpec.asInstanceOf[WriteFilesSpec]
-
+  override protected def doExecuteWrite(
+      writeFilesSpec: WriteFilesSpec): RDD[WriterCommitMessage] = {
```

Comment:
should `WriteFilesSpec` include less information as some information are already available in `WriteFilesExec`?

Reply:
Seems it's a bit hard. look at the current information:
```scala
case class WriteFilesSpec(
    description: WriteJobDescription,
    committer: FileCommitProtocol,
    concurrentOutputWriterSpecFunc: SparkPlan => Option[ConcurrentOutputWriterSpec])
```
- `ConcurrentOutputWriterSpec` and `FileCommitProtocol` contain the output spec so we can not replace them
- `WriteJobDescription` contains many information which includes what we pull out, but if we want to reduce something inside `WriteJobDescription`, we need to create a new class to hold others. I'm not sure it's worth to do that.

```scala
class WriteJobDescription(
    val uuid: String,
    val serializableHadoopConf: SerializableConfiguration,
    val outputWriterFactory: OutputWriterFactory,
    val allColumns: Seq[Attribute],
    val dataColumns: Seq[Attribute],
    val partitionColumns: Seq[Attribute],
    val bucketSpec: Option[WriterBucketSpec],
    val path: String,
    val customPartitionLocations: Map[TablePartitionSpec, String],
    val maxRecordsPerFile: Long,
    val timeZoneId: String,
    val statsTrackers: Seq[WriteJobStatsTracker])
```